File: src/main/scala/org/glavo/llvm/internal/JNIUtils.scala
package org.glavo.llvm.internal

import java.net.URL
import java.nio.file.Files

object JNIUtils {
  def Encoding: String = "UTF-8"

  val jvmName: String = System.getProperty("java.vm.name", "").toLowerCase

  val osName: String = System.getProperty("os.name", "").toLowerCase match {
    case s if jvmName.startsWith("dalvik") && s.startsWith("linux") =>
      "android"
    case s if s.startsWith("robovm") && s.startsWith("darwin") =>
      "ios"
    case s if s.startsWith("mac os x") || s.startsWith("darwin") =>
      "macosx"
    case s if s.contains("windows") =>
      "windows"
    case s =>
      val i = s.indexOf(' ')
      if (i > 0)
        s.substring(0, i)
      else
        s
  }

  val osArch: String = System.getProperty("os.arch", "").toLowerCase match {
    case "i386" | "i486" | "i586" | "i686" =>
      "x86"
    case "amd64" | "x86-64" | "x64" =>
      "x86_64"
    case s if s.startsWith("aarch64") || s.startsWith("armv8") || s.startsWith("arm64") =>
      "arm64"
    case s if s.startsWith("arm") ||
      System.getProperty("sun.arch.abi", "").toLowerCase == "gnueabihf"
      || System.getProperty("sun.boot.library.path", "").toLowerCase.contains("openjdk-armhf") =>
      "armhf"
    case s if s.startsWith("arm") || osName == "ios" =>
      "arm"
    case s => s
  }

  val platform: String = s"$osName-$osArch"

  val suffix: String = osName match {
    case "windows" => ".dll"
    case "macosx" => ".dylib"
    case _ => ".so"
  }

  val prefix: String = if (osName == "windows") "" else "lib"

  def load(url: URL): Unit = {
    val fn = url.toString.split("/").last

    val stream = url.openStream()
    val temp = Files.createTempFile("lib", "fn")
    Files.deleteIfExists(temp)
    Files.copy(stream, temp)
    stream.close()
    System.load(temp.toAbsolutePath.toFile.toString)
  }

  def load(cls: Class[_], libName: String): Unit = {
    load(cls.getResource(s"$platform/$prefix$libName$suffix"))
  }

  lazy val load: Unit = {
    load(getClass, "LLVM-5.0")
    load(getClass, "LTO")
    load(getClass, "llvm4s")
  }


  def str2bytes(str: String): Array[Byte]@ByteString = str.getBytes(Encoding)

  def bytes2str(bytes: Array[Byte]@ByteString): String = new String(bytes, Encoding)
}

File: internal/src/main/java/org/glavo/llvm/internal/JNIUtils.java
package org.glavo.llvm.internal;

import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Objects;

public final class JNIUtils {
    public static final String encoding = "UTF-8";

    private static final String PLATFORM;

    public static final String PREFIX;
    public static final String SUFFIX;


    static {
        String jvmName = System.getProperty("java.vm.name", "").toLowerCase();
        String osName = System.getProperty("os.name", "").toLowerCase();
        String osArch = System.getProperty("os.arch", "").toLowerCase();
        String abiType = System.getProperty("sun.arch.abi", "").toLowerCase();
        String libPath = System.getProperty("sun.boot.library.path", "").toLowerCase();
        if (jvmName.startsWith("dalvik") && osName.startsWith("linux")) {
            osName = "android";
        } else if (jvmName.startsWith("robovm") && osName.startsWith("darwin")) {
            osName = "ios";
            osArch = "arm";
        } else if (osName.startsWith("mac os x") || osName.startsWith("darwin")) {
            osName = "macosx";
        } else if (osName.contains("windows")) {
            osName = "windows";
        } else {
            int spaceIndex = osName.indexOf(' ');
            if (spaceIndex > 0) {
                osName = osName.substring(0, spaceIndex);
            }
        }
        if (osArch.equals("i386") || osArch.equals("i486") || osArch.equals("i586") || osArch.equals("i686")) {
            osArch = "x86";
        } else if (osArch.equals("amd64") || osArch.equals("x86-64") || osArch.equals("x64")) {
            osArch = "x86_64";
        } else if (osArch.startsWith("aarch64") || osArch.startsWith("armv8") || osArch.startsWith("arm64")) {
            osArch = "arm64";
        } else if ((osArch.startsWith("arm")) && ((abiType.equals("gnueabihf")) || (libPath.contains("openjdk-armhf")))) {
            osArch = "armhf";
        } else if (osArch.startsWith("arm")) {
            osArch = "arm";
        }

        PLATFORM = osName + "-" + osArch;

        switch (osName) {
            case "windows":
                PREFIX = "";
                SUFFIX = ".dll";
                break;
            case "macosx":
            case "ios":
                PREFIX = "lib";
                SUFFIX = ".dylib";
                break;
            default:
                PREFIX = "lib";
                SUFFIX = ".so";
        }
    }

    public static void load(URL url) {
        try (InputStream stream = url.openStream()) {
            Path temp = Files.createTempFile(PREFIX, SUFFIX);
            Files.deleteIfExists(temp);
            Files.copy(stream, temp);
            System.load(temp.toAbsolutePath().toFile().toString());
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(-1);
        }
    }

    private static boolean isLoaded = false;

    public static void load() {
        if (!isLoaded) {
            isLoaded = true;
            try {
                Class<?> cls = Class.forName(
                        String.format(
                                "%s.%s_%s",
                                NativeLibrary.class.getPackage().getName(),
                                NativeLibrary.class.getSimpleName(),
                                PLATFORM.replace('-', '_')
                        )
                );
                String[] libNames = (String[]) cls.getField("libNames").get(null);

                for (String libName : libNames) {
                    load(Objects.requireNonNull(cls.getResource(String.format("%s/%s%s%s", PLATFORM, PREFIX, libName, SUFFIX))));
                }

            } catch (Exception e) {
                e.printStackTrace();
                System.exit(-1);
            }
        }
    }

    public static String bytes2str(byte[] bytes) throws UnsupportedEncodingException {
        Objects.requireNonNull(bytes);

        return new String(bytes, encoding);
    }

    public static byte[] str2bytes(String str) throws UnsupportedEncodingException {
        Objects.requireNonNull(str);
        return str.getBytes(encoding);
    }
}
